src/backend/distributed/utils/metadata_cache.c

```diff
@@ -2570,7 +2590,16 @@ InitializeWorkerNodeCache(void)
 	}
 
 	/* now, safe to destroy the old hash */
-	hash_destroy(oldWorkerNodeHash);
+	hash_destroy(WorkerNodeHash);
+
+	if (WorkerNodeArray != NULL)
+	{
+		pfree(WorkerNodeArray);
+	}
+
+	WorkerNodeCount = newWorkerNodeCount;
+	WorkerNodeArray = newWorkerNodeArray;
+	WorkerNodeHash = newWorkerNodeHash;
```

Comment:
If we were in a multi-threaded context this would be dangerous but I suppose the chance of us ever using threads is 0% so... let's ship it!